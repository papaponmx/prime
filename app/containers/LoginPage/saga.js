import { takeLatest, put } from 'redux-saga/effects';
import { LOGIN_USER } from './constants';
import { loginUserSuccess, loginUserError } from './actions';
import { app, googleProvider } from '../../configureStore';

export function* loginUserSaga() {
  let token;
  let userInformation;
  try {
    yield app
      .auth()
      .signInWithPopup(googleProvider)
      .then(result => {
        // IDEA: Encript token
        token = result.credential.accessToken;
        userInformation = JSON.stringify(result.user);
        localStorage.setItem('prime-app-UserToken', token);
        localStorage.setItem('prime-app-UserInformation', userInformation);
      });
    yield put(
      loginUserSuccess({
        token,
        userInformation,
      }),
    );
  } catch (error) {
    loginUserError({
      code: error.code,
      message: error.message,
    });
  }
}

// Individual exports for testing
export default function* watcherSaga() {
  yield takeLatest(LOGIN_USER, loginUserSaga);
  // See example in containers/HomePage/saga.js
}
