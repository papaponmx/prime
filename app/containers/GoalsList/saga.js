// import firebase from 'firebase/app';
import { takeLatest, put, select } from 'redux-saga/effects';
import { FETCH_GOALS } from './constants';
import { fetchGoalsSuccess, fetchGoalsError } from './actions';
import { database as db } from '../../configureStore';
import { selectUserUid } from '../LoginPage/selectors';

export function* fetchGoals() {
  let list;
  const uid = yield select(selectUserUid);
  yield db.settings({ timestampsInSnapshots: true });

  try {
    yield db
      .collection('users')
      .doc(uid)
      .collection('goals')
      .get()
      .then(res => {
        list = res.docs.map(goal => goal.data());
        return list;
      });
  } catch (error) {
    yield put(fetchGoalsError(error));
  }
  yield put(fetchGoalsSuccess(list));
}

// Individual exports for testing
export default function* defaultSaga() {
  yield takeLatest(FETCH_GOALS, fetchGoals);

  // See example in containers/HomePage/saga.js
}
