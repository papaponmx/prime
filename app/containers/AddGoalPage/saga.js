// import firebase from 'firebase/app';
import { takeLatest, put, select } from 'redux-saga/effects';
import { push } from 'react-router-redux';
import { database as db } from '../../configureStore';
import { selectUserUid } from '../LoginPage/selectors';
import uniqueId from '../../utils/uniqueId';
import { ADD_GOAL } from '../GoalsList/constants';

export function* addGoalSaga({ payload }) {
  console.log('DQTP, inside addGoalSaga');

  const createdAt = new Date().getTime();
  const id = uniqueId('goal');
  const { goal: name, dueDate: date } = payload;
  const d = new Date(date);
  const dueDate = d.getTime();
  // FIXME: Transform this in a toMilis Function
  const uid = yield select(selectUserUid);
  yield db
    .collection('users')
    .doc(uid)
    .collection('goals')
    .doc(id)
    .set({
      createdAt,
      dueDate,
      id,
      isDone: false,
      name,
    })
    .then(() => '')
    .catch(error => error);

  yield put(push('/'));
  // TODO: Add Goal to state
}

// Individual exports for testing
export default function* defaultSaga() {
  yield takeLatest(ADD_GOAL, addGoalSaga);
}
