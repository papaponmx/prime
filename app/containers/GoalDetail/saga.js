import { put, takeLatest, select } from 'redux-saga/effects';
import { push } from 'react-router-redux';
import { DELETE_GOAL, MARK_GOAL_AS_DONE } from '../GoalsList/constants';
import { database as db, app as firebase } from '../../configureStore';
import { deleteGoalError } from '../GoalsList/actions';
import { selectGoalById } from '../GoalsList/selectors';

// Individual exports for testing

export function* deleteGoalSaga({ payload: id }) {
  try {
    yield db
      .collection('users')
      .doc(firebase.auth().getUid())
      .collection('goals')
      .doc(id)
      .delete();

    yield put(push('/'));
  } catch (error) {
    yield put(deleteGoalError(error));
  }
  // TODO: Delete Goal From state and handle error
}

export function* updateGoalSaga({ type, payload: id }) {
  let data = {};
  if (type === MARK_GOAL_AS_DONE) {
    data = {
      isDone: true,
    };
  } else {
    // TODO: Test and define this logic
    yield select(selectGoalById(id));
  }

  // TODO: This  update is not working
  try {
    yield db
      .collection('users')
      .doc(firebase.auth().getUid())
      .collection('goals')
      .doc(id)
      .update(data);
    yield put(push('/'));
    // TODO: Update goal on state
  } catch (error) {
    // console.log(error);
  }
  // TODO: Handle error
}

export default function* defaultSaga() {
  // See example in containers/HomePage/saga.js
  yield takeLatest(DELETE_GOAL, deleteGoalSaga);
  yield takeLatest(MARK_GOAL_AS_DONE, updateGoalSaga);
}
