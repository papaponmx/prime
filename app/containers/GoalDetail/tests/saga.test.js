// import { push } from 'react-router-redux';
// import { put } from 'redux-saga/effects';
import { deleteGoalSaga } from '../saga';

/**
 * Test sagas
 */

/* eslint-disable redux-saga/yield-effects */
// import { defaultSaga } from '../saga';

// const generator = defaultSaga();

describe('defaultSaga Saga', () => {
  let deleteGoalGenerator;

  beforeEach(() => {
    deleteGoalGenerator = deleteGoalSaga({ payload: '123' });

    const selectDescriptor = deleteGoalGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = deleteGoalGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  xit('Should dispatch deleteGoalSuccess action if request is successful', () => {});

  xit('should dispatch the push(`/`) action if it request is successful', () => {
    const putDescriptor = deleteGoalGenerator.next().value;
    expect(JSON.stringify(putDescriptor, null, 2)).toMatch(`put(push('/'))`);
  });

  xit('Expect to have unit tests specified', () => {
    expect(true).toEqual(false);
  });
});
