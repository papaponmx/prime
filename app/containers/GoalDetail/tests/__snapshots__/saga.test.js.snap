// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultSaga Saga should dispatch the push(\`/\`) action if it request is successful 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": [FirebaseError: Function CollectionReference.doc() requires its first argument to be of type string, but it was: null],
      "type": "app/GoalsList/DELETE_GOAL_ERROR",
    },
    "channel": null,
  },
}
`;

exports[`defaultSaga Saga should dispatch the push(\`/\`) action if it request is successful 2`] = `undefined`;
